### PR TITLE
#206: Implemented UninstallCommandlet

### LINKED ISSUE #206: Implement UninstallCommandlet
### Expected behavior

An UninstallCommandlet is implemented to efficiently uninstall commandlet tools.

### Actual behavior
Currently, there's no opportunity to uninstall tools through devonfw within the IDEasy project. The only workaround available is manually deleting the tool folders from the project structure.

UnistallCommandlet  is supposed to delegate to the uninstall method of LocalToolCommandlet.

### PR DESCRIPTION
**Pull request by ndemirca**
Closes #206

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/commandlet/UninstallCommandlet.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    String commandletName = this.tool.getValue().getName();
+    Path softwarePath = context.getSoftwarePath().resolve(commandletName);
+    if (Files.exists(softwarePath)) {
+      FileAccess fileAccess = context.getFileAccess();
+      try {
+        fileAccess.delete(softwarePath);
+        this.context.success("Successfully uninstalled " + commandletName);
+      } catch (Exception e) {
+        throw new IllegalStateException("Couldn't uninstall " + commandletName);
```
Comment (mvomiero) [inline cli/src/main/java/com/devonfw/tools/ide/commandlet/UninstallCommandlet.java:47]: Maybe you have to rethrow the original exception in order to preserve the stacktrace, as in the coding conventions?
`throw new IllegalStateException("Couldn't uninstall " + commandletName, e);`

https://github.com/devonfw/IDEasy/blob/main/documentation/coding-conventions.adoc#catching-and-handling-exceptions

File: cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    // arrange
+    String toolName = "npm";
+    IdeTestContext context = newContext(PROJECT_BASIC);
+    UninstallCommandlet uninstallCommandlet = context.getCommandletManager().getCommandlet(UninstallCommandlet.class);
+    uninstallCommandlet.tool.setValueAsString(toolName, context);
+    // act
+    uninstallCommandlet.run();
+    // assert
+    Path softwarePath = context.getSoftwarePath();
```
Comment (mvomiero) [inline cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java:37]: The variable is unused

File: cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    // arrange
+    String toolName = "eclipse";
+    IdeTestContext context = newContext(PROJECT_BASIC);
+    UninstallCommandlet uninstallCommandlet = context.getCommandletManager().getCommandlet(UninstallCommandlet.class);
+    uninstallCommandlet.tool.setValueAsString(toolName, context);
+    // act
+    uninstallCommandlet.run();
+    // assert
+    Path softwarePath = context.getSoftwarePath();
```
Comment (mvomiero) [inline cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java:53]: the variable is unused

**Review by mvomiero — CHANGES_REQUESTED** [review]
Nice implementation for this commandlet! 👍   just some small fixes suggested, other then that LGTM

**Comment by salimbouch** [comment]
Just a suggestion, but maybe you could delegate the uninstall process to LocalToolCommandlet just like you described in #206, it seems cleaner.

File: cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java (hunk 2 above)
Comment (ndemirca) [reply cli/src/test/java/com/devonfw/tools/ide/commandlet/UninstallCommandletTest.java]: It was test variable, sorry!

**Review by mvomiero — APPROVED** [review]

**Review by hohwille — APPROVED** [review]
@ndemirca Thanks for your PR. All looking good and ready for merge 👍 
FYI: As nice to have remark, the methods like `testUninstallCommandletRun_WithExistingCommandlet()` are not as intuitive as the commandlets always exist. A more compact and self-explaining name could be `testUninstallWithAlreadyInstalledTool` (and `testUninstallWithoutInstalledTool`).

But nothing wrong here. Thanks for your nice work 👍

**Pull request merged**